src/coreclr/jit/valuenum.cpp

```diff
@@ -7768,82 +7777,80 @@ void Compiler::fgValueNumberAssignment(GenTreeOp* tree)
                             assert(fldSeq != nullptr);
                         }
 
-                        // Get a field sequence for just the first field in the sequence
-                        //
-                        FieldSeqNode* firstFieldOnly = GetFieldSeqStore()->CreateSingleton(fldSeq->m_fieldHnd);
+                        // The value number from the rhs of the assignment
+                        ValueNum storeVal  = rhsVNPair.GetLiberal();
+                        ValueNum newHeapVN = ValueNumStore::NoVN;
 
-                        // The final field in the sequence will need to match the 'indType'
+                        // We will check that the final field in the sequence matches 'indType'.
                         var_types indType = lhs->TypeGet();
-                        ValueNum  fldMapVN =
-                            vnStore->VNApplySelectors(VNK_Liberal, fgCurMemoryVN[GcHeap], firstFieldOnly);
-
-                        // The type of the field is "struct" if there are more fields in the sequence,
-                        // otherwise it is the type returned from VNApplySelectors above.
-                        var_types firstFieldType = vnStore->TypeOfVN(fldMapVN);
-
-                        // The value number from the rhs of the assignment
-                        ValueNum storeVal    = rhsVNPair.GetLiberal();
-                        ValueNum newFldMapVN = ValueNumStore::NoVN;
 
                         // when (obj != nullptr) we have an instance field, otherwise a static field
                         // when (staticOffset != nullptr) it represents a offset into a static or the call to
                         // Shared Static Base
                         if ((obj != nullptr) || (staticOffset != nullptr))
                         {
-                            ValueNum valAtAddr = fldMapVN;
-                            ValueNum normVal   = ValueNumStore::NoVN;
+                            var_types firstFieldType;
+                            ValueNum  firstFieldSelectorVN =
+                                vnStore->VNForFieldSelector(fldSeq->GetFieldHandle(), &firstFieldType);
 
+                            // Construct the "field map" VN. It represents memory state of the first field
+                            // of all objects on the heap. This is our primary map.
+                            ValueNum fldMapVN = vnStore->VNForMapSelect(VNK_Liberal, firstFieldType,
+                                                                        fgCurMemoryVN[GcHeap], firstFieldSelectorVN);
```

Comment:
Should we be using `TYP_REF` instead of `firstFieldType` here accordingly to the discussion last week?

Reply:
Yes we should, but as standalone change, it will create regressions because the reading side of this code needs to be updated to use the same type (it is using `firstFieldType` at the moment). This is the reason for the 4th commit:
```
This is a temporary change to minimize diffs.
The issue is that we need to update both the
reading and writing code at once if we want
to change the types of the maps without some
regressions due to new artificial mismatches.
```
In the next change, I will update the reading side, and in the change after that - the types for the heaps/"first field maps" (then will come the many changes needed to enable clean validation, after that the validation itself, and finally the actual bugfixes).